From af71d660b54edde6ae30380e5c378e60f68b3ce8 Mon Sep 17 00:00:00 2001
From: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
Date: Thu, 3 Jul 2025 12:47:17 +0300
Subject: [PATCH 104/108] [WIP-FRL] drm/connector: hdmi: Handle FRL in
 hdmi_clock_valid()

Do not limit clock validation to max_tmds_clock if the sink advertises
FRL support.

TODO: extend FRL capability verification at source (driver) level.

Signed-off-by: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
---
 drivers/gpu/drm/display/drm_hdmi_state_helper.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/display/drm_hdmi_state_helper.c b/drivers/gpu/drm/display/drm_hdmi_state_helper.c
index a561f124be99..0608443985f1 100644
--- a/drivers/gpu/drm/display/drm_hdmi_state_helper.c
+++ b/drivers/gpu/drm/display/drm_hdmi_state_helper.c
@@ -535,7 +535,9 @@ hdmi_clock_valid(const struct drm_connector *connector,
 	const struct drm_connector_hdmi_funcs *funcs = connector->hdmi.funcs;
 	const struct drm_display_info *info = &connector->display_info;
 
-	if (info->max_tmds_clock && clock > info->max_tmds_clock * 1000)
+	//TODO: add proper FRL rate check
+	if ((!info->hdmi.max_frl_rate_per_lane || !info->hdmi.max_lanes) &&
+	    (info->max_tmds_clock && clock > info->max_tmds_clock * 1000))
 		return MODE_CLOCK_HIGH;
 
 	if (funcs && funcs->tmds_char_rate_valid) {
-- 
2.34.1

